pydantic_ai_slim/pydantic_ai/ext/temporal/_model.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import annotations
+
+from collections.abc import AsyncIterator
+from contextlib import asynccontextmanager
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Any, Callable
+
+from pydantic import ConfigDict, with_config
+from temporalio import activity, workflow
+from temporalio.workflow import ActivityConfig
+
+from pydantic_ai._run_context import RunContext
+from pydantic_ai.agent import EventStreamHandler
+from pydantic_ai.exceptions import UserError
+from pydantic_ai.messages import (
+    ModelMessage,
+    ModelResponse,
+    ModelResponseStreamEvent,
+)
+from pydantic_ai.models import Model, ModelRequestParameters, StreamedResponse
+from pydantic_ai.models.wrapper import WrapperModel
+from pydantic_ai.settings import ModelSettings
+from pydantic_ai.usage import Usage
+
+from ._run_context import TemporalRunContext
+
+
+@dataclass
+@with_config(ConfigDict(arbitrary_types_allowed=True))
+class _RequestParams:
+    messages: list[ModelMessage]
+    model_settings: ModelSettings | None
+    model_request_parameters: ModelRequestParameters
+    serialized_run_context: Any
+
+
+class _TemporalStreamedResponse(StreamedResponse):
+    def __init__(self, model_request_parameters: ModelRequestParameters, response: ModelResponse):
+        super().__init__(model_request_parameters)
+        self.response = response
+
+    async def _get_event_iterator(self) -> AsyncIterator[ModelResponseStreamEvent]:
+        return
+        # noinspection PyUnreachableCode
+        yield
+
+    def get(self) -> ModelResponse:
+        """Build a [`ModelResponse`][pydantic_ai.messages.ModelResponse] from the data received from the stream so far."""
+        return self.response
+
+    def usage(self) -> Usage:
+        """Get the usage of the response so far. This will not be the final usage until the stream is exhausted."""
+        return self.response.usage
+
+    @property
+    def model_name(self) -> str:
+        """Get the model name of the response."""
+        return self.response.model_name or ''
+
+    @property
+    def timestamp(self) -> datetime:
+        """Get the timestamp of the response."""
+        return self.response.timestamp
+
+
+class TemporalModel(WrapperModel):
+    def __init__(
+        self,
+        model: Model,
+        event_stream_handler: EventStreamHandler[Any],
+        activity_config: ActivityConfig = {},
+        run_context_type: type[TemporalRunContext] = TemporalRunContext,
+    ):
+        super().__init__(model)
+        self.activity_config = activity_config
+        self.event_stream_handler = event_stream_handler
+        self.run_context_type = run_context_type
+
+        id = '_'.join([model.system, model.model_name])
+
+        @activity.defn(name=f'model__{id}__request')
```

Comment:
This should have the agent name in it so multiple agents can be used in a workflow

Reply:
Also, this means we need to ensure that agent names are unique right? Does that happen "eagerly"?